[Quantization] Validate checkpointing works in mix mode quant

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111923

Summary:
Adding a test to ensure that when training a model with qat and applying
ptq later is still compatible with checkpointing

Test Plan:
Test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50602861](https://our.internmc.facebook.com/intern/diff/D50602861)

kimishpatel committed:
[Quantization] Validate checkpointing works in mix mode quant

Summary:
Adding a test to ensure that when training a model with qat and applying
ptq later is still compatible with checkpointing

Test Plan:
Test added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50602861).

kimishpatel committed:
Update on "[Quantization] Validate checkpointing works in mix mode quant"


Summary:
Adding a test to ensure that when training a model with qat and applying
ptq later is still compatible with checkpointing

Test Plan:
Test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50602861](https://our.internmc.facebook.com/intern/diff/D50602861)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50602861).

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -732,3 +732,78 @@ def test_mixing_qat_ptq(self):
+        state_dict_before = copy.deepcopy(model.state_dict())
+        torch.save(model.state_dict(), buffer)
+        buffer.seek(0)
+        model.load_state_dict(torch.load(buffer))
```
nit: to simulate the real use case, this model should be initialized from the beginning I think

kimishpatel replied:
what do you mean?

jerryzh168 replied:
oh I mean we should repeat L741-744 here before loading the state_dict to simulate the use case that people do checkpointing, and resume training a bit later

kimishpatel replied:
Thanks for pointing this out. It does require running through prepare_qat step again.

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -732,3 +732,78 @@ def test_mixing_qat_ptq(self):
+        quantizer.set_global(quantization_config)
+        model_pt2e = prepare_pt2e(model_pt2e, quantizer)
+        after_prepare_result_pt2e = model_pt2e(*example_inputs)
+        model_pt2e = convert_pt2e(model_pt2e)
```
fold_quantize=True, maybe we should prioritize the BE work to remove this flag?

kimishpatel replied:
Thanks. Keep forgetting

kimishpatel committed:
Update on "[Quantization] Validate checkpointing works in mix mode quant"


Summary:
Adding a test to ensure that when training a model with qat and applying
ptq later is still compatible with checkpointing

Test Plan:
Test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50602861](https://our.internmc.facebook.com/intern/diff/D50602861)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50602861).

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -720,7 +720,88 @@ def test_mixing_qat_ptq(self):
+            output.backward()
+            optimizer.step()
+
+        state_dict_before = copy.deepcopy(model.state_dict())
```
nit: could you also add some comments to specify the context for each section of code? like QAT, checkpoint, load checkpoint, PTQ etc.
